src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs

```diff
@@ -196,6 +196,33 @@ public async ValueTask<IReadOnlyCollection<VersionInfoContextInfo>> GetPackageVe
             return await GetPackageVersionsAsync(identity, packageSources, includePrerelease, isTransitive, projects: null, cancellationToken);
         }
 
+        public async Task<(bool?, string)> TryGetPackageReadMeAsync(PackageIdentity identity, IReadOnlyCollection<PackageSourceContextInfo> packageSources, bool includePrerelease, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(identity);
+            Assumes.NotNullOrEmpty(packageSources);
+
+            IReadOnlyCollection<SourceRepository> sourceRepositories = await _sharedServiceState.GetRepositoriesAsync(packageSources, cancellationToken);
+            var globalPackage = await _globalPackageFolderRepositoriesLazy.GetValueAsync(cancellationToken);
+            var localPackageFolder = await _packagesFolderLocalRepositoryLazy.GetValueAsync(cancellationToken);
+
+            var localReadmeResources = sourceRepositories
+                .Concat(globalPackage)
+                .Concat(new List<SourceRepository> { localPackageFolder })
+                .Select(x => x.GetResource<ReadmeResource>(cancellationToken))
+                .Where(x => x is LocalReadmeResource);
```

Comment:
Code using `SourceRepository` should not need to type cast to feed specific types. The point of it is to abstract away those implementation details so that local feeds and http feeds have the same API. This points to a flawed API design.

Reply:
What I wanted to do here is to determine which Source Repository to use based on what version of the "ReadmeResource" is available on it. I could just take first readme that's returned from the available sources. I can just return whichever gives a readme first I suppose